cmd/habitat/community/state/transitions.go

```diff
@@ -211,3 +223,158 @@ func (t *AddNodeTransition) Validate(oldState *community.CommunityState) error {
 
 	return nil
 }
+
+type StartProcessTransition struct {
+	Process *community.Process
+}
+
+func (t *StartProcessTransition) Type() string {
+	return TransitionTypeStartProcess
+}
+
+func (t *StartProcessTransition) Patch(oldState *community.CommunityState) ([]byte, error) {
+	marshaledProcess, err := json.Marshal(t.Process)
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(fmt.Sprintf(`[{
+		"op": "add",
+		"path": "/processes/-",
+		"value": %s
+	}]`, string(marshaledProcess))), nil
+}
+
+func (t *StartProcessTransition) Validate(oldState *community.CommunityState) error {
+	// make sure no process with same process id and app name is already started
+	for _, p := range oldState.Processes {
+		if p.ID == t.Process.ID {
+			return fmt.Errorf("process ID %s already used in community %s", t.Process.ID, oldState.CommunityID)
+		}
+		if p.AppName == t.Process.AppName {
+			return fmt.Errorf("community %s already has app %s running", oldState.CommunityID, t.Process.ID)
+		}
+	}
+
+	return nil
+}
+
+type StopProcessTransition struct {
+	ProcessID string
+}
+
+func (t *StopProcessTransition) Type() string {
+	return TransitionTypeStopProcess
+}
+
+func (t *StopProcessTransition) Patch(oldState *community.CommunityState) ([]byte, error) {
+	for i, p := range oldState.Processes {
+		if p.ID == t.ProcessID {
+			return []byte(fmt.Sprintf(`[{
+				"op": "remove",
+				"path": "/processes/%d"
+			}]`, i)), nil
+		}
+	}
+
+	return nil, fmt.Errorf("process ID %s not found in processes list", t.ProcessID)
+}
+
+func (t *StopProcessTransition) Validate(oldState *community.CommunityState) error {
+	// ensure no process_instances with the process ID are still around
+	for _, i := range oldState.ProcessInstances {
```

Comment:
wait i thought `Validate` is called before the transitions is applied to make sure it is safe? so wouldn't we want to end the process with the same id, not return an error?

is `Validate` called after the transition? because then i think the one for `StartProcess` doesn't make sense

Reply:
Here we are checking that there are no `ProcessInstances` still running that are mapped to the `Process` we are trying to stop. If the controller proposing the transition is responsible for stopping all `ProcessInstances` before stopping the `Process` itself. The validation is run before the transition. I think its better for us to return errors when a controller uses a transition incorrectly, instead of trying to "correct" the mistake in the transition logic